build/src/lib.rs

```diff
@@ -148,52 +286,59 @@ pub struct BinaryBuilder {
     /// - If they are not found (i. e. a dev version of nlprule is used) and `fallback_to_build_dir`is true
     /// downloads the latest build directory  and builds the binaries from it.
     /// This can still fail if the dev version is sufficiently outdated for the latest build dir.
-    /// In that case, the user can't be helped.
-    fn build_language(&mut self, lang_code: &str) {
-        let tokenizer_out = self.out_dir.join(tokenizer_filename(lang_code));
-        let rules_out = self.out_dir.join(rules_filename(lang_code));
+    /// In that case, the user is encouraged to update to a release or a newer git sha.
+    fn build_language(&mut self, lang_code: &str) -> Result<()> {
+        // adjust the destination path for now
+        let path_transform = |out: PathBuf| -> Result<PathBuf> {
+            Ok(if let Some(ref transform_path_fn) = self.transform_path_fn {
+                transform_path_fn(out).map_err(Error::TransformError)?
+            } else {
+                out
+            })
+        };
+
+        let tokenizer_out = path_transform(self.out_dir.join(tokenizer_filename(lang_code)))?;
+        let rules_out = path_transform(self.out_dir.join(rules_filename(lang_code)))?;
 
         let mut did_not_find_binaries = false;
 
-        for (binary, out) in [
+        for (binary, out) in &[
             (Binary::Tokenizer, &tokenizer_out),
             (Binary::Rules, &rules_out),
-        ]
-        .iter()
-        {
-            let response = get_binary(&self.version, lang_code, *binary, self.cache_dir.as_ref());
-
-            (match response {
-                Ok(mut reader) => {
-                    let mut bytes = Vec::new();
-                    reader
-                        .read_to_end(&mut bytes)
-                        .expect("reading binary bytes failed");
-
-                    fs::write(out, bytes).expect("writing binary to file failed");
-
-                    Ok(())
-                }
-                Err(Error::RequestError(error)) => {
-                    if let Some(404) = error.status().map(|x| x.as_u16()) {
-                        // binaries for this version are not distributed, fall back to building them
-                        did_not_find_binaries = true;
-                        break;
-                    }
-
-                    Err(error.into())
+        ] {
+            let out = out.to_owned().to_owned();
+            if let Err(e) = assure_binary_availability(
+                &self.version,
+                lang_code,
+                *binary,
+                self.cache_dir.as_ref(),
+                self.transform_path_fn.as_ref().map(|x| x.as_ref()),
+                self.transform_data_fn.as_ref().map(|x| x.as_ref()),
+                out,
+            ) {
+                if let Error::BinariesNotFound = e {
+                    did_not_find_binaries = true;
+                    break;
                 }
```

Comment:
This silently ignores any error happening in `assure_binary_availability` besides `Error::BinariesNotFound`. Should be propagated.

Reply:
Good catch!